### ISSUE TITLE
Corrupted buffer in Chrome 79 (legacy web engines)

### ISSUE THREAD

**Issue by dmitrylyzo** [ThaUnknown/jassub#28]
After https://github.com/jellyfin/jellyfin-web/pull/4613 (+https://github.com/jellyfin/jellyfin-web/pull/4702), JASSUB stopped working in Chrome 79 (and presumably on TVs with legacy web engines).

1.7.1
![jassub-171](https://github.com/ThaUnknown/jassub/assets/56478732/cd23e8c5-8aa2-4146-a1f1-6efea033d351)
It looks like the buffer was copied incorrectly. It is updated when the event changes, but remains corrupted.
With `asyncRender: false`, it seems to work.

1.5.13 (for reference)
![jassub-1513](https://github.com/ThaUnknown/jassub/assets/56478732/23ee7ac0-b6a6-4d89-b1b0-787de50ed860)

**Sample**
[test.ass.txt](https://github.com/ThaUnknown/jassub/files/12614246/test.ass.txt)

**Comment by ThaUnknown** [comment]
hmmmm this is a nasty one

TLDR: I recommend you disable async render based on user agent if UA is < Chrome79

I'm not really sure what happens here, while async render is running, the underlying data won't ever be mutated, so it's not like the corruption is caused by something changing the data, so I assume this is a bug in chromium that was fixed a while back

this most likely stems from the fact JASSUB passes memory references instead of copies to the browser, which causes this, simply because it's a LOOOOOOT faster than copies

I could fix this in JASSUB, but I'm not sure I'd want to
- I could fix it by copying memory, but I don't like that idea because it's pretty much saying "hey, all clients need to render stuff 10% slower [random value, but it's likely higher]" just because an older engine has bug which no longer exists in newer version
- I could parse useragent, but that's also a bad idea in case someone wants to run the library in NodeJS using node-canvas

I really think it'd be better to add a check in the jellyfin-web app, in the constructor simply as you said disable asyncRender if it's an old version of webOS disable asyncRender, as there isn't much performance to gain by using it on those older engines as the hardware running it likely won't be able to render any complex typesetting even with the help of asyncRender

I'll see if I can make a PR sometime next week if you aren't sure how to do this

**Comment by dmitrylyzo** [comment]
`asyncRender: true` works with 1.5.13 (I tried reversing those PRs).

**Comment by ThaUnknown** [comment]
and it's also significantly slower on heavy TS

using an older version means you lose quite a few important features, mainly Firefox performance improvements, SIMD improvements and colorspace conversion

**Comment by dmitrylyzo** [comment]
Found that replacing `subarray` with `slice` fixes the corruption. But it creates a copy, iiuc.
https://github.com/ThaUnknown/jassub/blob/9476f4f9a0f226a23920bea7f91be467f45c3195/src/worker.js#L236

Introduced https://github.com/ThaUnknown/jassub/commit/41d630460e9bd49ab7a99193babecc8a3147965e#diff-a19812fe5175f5ae8fccdf2c9400b66ea4408f519c4208fded5ae4c3365cac4dR235

**Comment by ThaUnknown** [comment]
yes as I said it's because of the references not copies, but again, changing this drops performance a LOT, so it's way better to parse useragent and disable asyncRender on old browsers which have bugs in createImageBitmap, rather than reducing the rendering performance of all browsers that don't even have the issue

*realisation*

hm I'll try reproducting the bug, and making a test for it like it's done for the alpha bug, but I'm very afraid its a WASM thing

then simply slice instead of subarray on an if check if the bug exists

**Comment by ThaUnknown** [comment]
I managed to track down and reproduce the bug, TLDR chromium used to ignore the offset from a subarray completly, and also mangle some of the data [massive woes], so i'll add a check for this, and set async render to use slice if this happens

**Comment by ThaUnknown** [comment]
this should be fixed in 1.7.2

**Comment by dmitrylyzo** [comment]
> also any chance i could get the video that you used? seems cool for debugging timestamps

Nothing special.
```sh
#!/bin/bash

IMAGE=banner-light.png
OUTPUT_IMAGE=out.png
OUTPUT=out.mp4

H264LEVEL=5.1

convert "$IMAGE" -background gray -gravity center -extent 1920x1080 "$OUTPUT_IMAGE"
ffmpeg -loop 1 -i "$OUTPUT_IMAGE" -c:v libx264 -profile:v high -level:v $H264LEVEL -t 300 -pix_fmt yuv420p -vf scale=1920x1080 -vf "drawtext=text='%{pts\:hms}': x=(w-tw)/2: y=h-(2*lh): fontcolor=white: fontsize=48" "$OUTPUT"
```
You may need to set the I-frame interval, because it is 10s due to the still image.

**Comment by ThaUnknown** [comment]
fixed the wrong function location in 1.7.3

**Comment by ThaUnknown** [comment]
@dmitrylyzo I assume this fixed it?

**Comment by dmitrylyzo** [comment]
You reverted the last change too much and [this comment](https://github.com/ThaUnknown/jassub/commit/03da58662ed58ddbaa34f511ee9b0f1a9d043690#r127541599) is valid again.
